core/src/main/java/org/apache/iceberg/rest/auth/JWT.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.rest.auth;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import java.time.Instant;
+import java.util.Base64;
+import java.util.Optional;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.JsonUtil;
+import org.immutables.value.Value;
+
+@Value.Immutable
+public interface JWT {
+
+  String token();
+
+  long expiresAtEpochMillis();
+
+  default long expiresInMillis() {
+    return expiresAtEpochMillis() - Instant.now().toEpochMilli();
+  }
+
+  default boolean expiredAt(long expiredAtEpochMillis) {
+    return expiresAtEpochMillis() < expiredAtEpochMillis;
+  }
+
+  static Optional<JWT> of(String token) {
+    try {
+      Preconditions.checkArgument(null != token, "Invalid JWT: null");
+      String[] parts = token.split("\\.");
+      Preconditions.checkArgument(parts.length == 3, "Invalid JWT: %s", token);
+
+      // Parse the payload JSON
+      JsonNode jsonNode = JsonUtil.mapper().readTree(Base64.getUrlDecoder().decode(parts[1]));
+      Long expiresAt = JsonUtil.getLongOrNull("exp", jsonNode);
```

Comment:
Do we need a class for JWT, or could this be a util method in `OAuth2Util`?

If it were a method in the util class, then we could support more token types or patterns (not just `exp`) in the future.

Reply:
I initially had this inside a method but the downside was basically that you'd need to parse the token multiple times if you'd wanted to access token expiration multiple times

Reply:
Those are both in the same method. Why not just use a local variable?

Reply:
sorry I'm not sure I'm following/understanding your argument properly. Could you elaborate here please what kind of change you'd like to see here?